File: packages/xpute/runtime/src/clock.ts
// @xpute/runtime/clock.ts

/**
 * The host's clock, in milliseconds: `performance.now()` unless something
 * installs another — a test that has to hold time still.
 */

import type { f64 } from "@xpute/core/abi/word.ts";

let clock: () => f64 = () => performance.now();

/** Installs the clock. */
export function set_clock(next: () => f64): void {
  clock = next;
}

/** The host's time, in milliseconds. */
export function now(): f64 {
  return clock();
}

File: crates/xpute/runtime/src/clock.rs
// xpute-runtime/clock.rs

//! The host's clock, in milliseconds: what `performance.now()` is to a
//! script. The host installs it before the first turn; until then time
//! stands at zero, which is what a native test that never installs one runs
//! on.

use core::cell::Cell;

thread_local! {
    static CLOCK: Cell<Option<fn() -> f64>> = const { Cell::new(None) };
}

/// Installs the host's clock.
pub fn set_clock(clock: fn() -> f64) {
    CLOCK.with(|c| c.set(Some(clock)));
}

/// The host's time, in milliseconds.
pub fn now() -> f64 {
    CLOCK.with(|c| c.get()).map_or(0.0, |f| f())
}
